src/main/java/org/apache/sysds/runtime/instructions/fed/UnaryMatrixFEDInstruction.java

```diff
@@ -63,17 +72,316 @@ public static UnaryMatrixFEDInstruction parseInstruction(String str) {
 	
 	@Override 
 	public void processInstruction(ExecutionContext ec) {
+		if(getOpcode().startsWith("ucum"))
+			processCumulativeInstruction(ec);
+		else {
+			MatrixObject mo1 = ec.getMatrixObject(input1);
+
+			//federated execution on arbitrary row/column partitions
+			//(only assumption for sparse-unsafe: fed mapping covers entire matrix)
+			FederatedRequest fr1 = FederationUtils.callInstruction(instString,
```

Comment:
The overall structure of the federated instruction requires a rework - I think I understand that you wanted to cover both row and column partitioned matrices with one code-path, but this way we currently perform much more computation than necessary. Also, the data is pulled into the coordinator which defeats its federated execution.

Reply:
for example for a cumsum on row-partitioned data, it seems this PR performs a ucumk+, pulls data to the coordinator, slices, out the last row, aggregates offsets, and then broadcasts these offsets and adds them to the matrix.

Instead, we should do a colSums per federated partition, obtains these aggregates, rbind them, local cumsum, and then broadcasts, add and perform the actual cumsum.